spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaConsumerBackoffManager.java

```diff
@@ -47,87 +53,162 @@
 
+	private static final LogAccessor logger = new LogAccessor(LogFactory.getLog(RetryTopicConfigurer.class));
 	/**
 	 * Internal Back Off Clock Bean Name.
 	 */
 	public static final String INTERNAL_BACKOFF_CLOCK_BEAN_NAME = "internalBackOffClock";
 
+	private static final int TIMING_CORRECTION_THRESHOLD = 100;
+
+	private static final int POLL_TIMEOUTS_FOR_CORRECTION_WINDOW = 2;
+
 	private final ListenerContainerRegistry registry;
 
-	private final Map<TopicPartition, Context> backOffTimes;
+	private final Map<TopicPartition, Context> backOffContexts;
 
 	private final Clock clock;
 
+	private final TaskExecutor taskExecutor;
+
+	private final Sleeper sleeper;
+
 	public KafkaConsumerBackoffManager(ListenerContainerRegistry registry,
-									@Qualifier(INTERNAL_BACKOFF_CLOCK_BEAN_NAME) Clock clock) {
+									@Qualifier(INTERNAL_BACKOFF_CLOCK_BEAN_NAME) Clock clock,
+									TaskExecutor taskExecutor,
+									Sleeper sleeper) {
 
 		this.registry = registry;
 		this.clock = clock;
-		this.backOffTimes = new HashMap<>();
+		this.taskExecutor = taskExecutor;
+		this.sleeper = sleeper;
+		this.backOffContexts = new HashMap<>();
 	}
 
 	/**
 	 * Backs off if the current time is before the dueTimestamp provided
 	 * in the {@link Context} object.
-	 * @param context the state that will be used for backing off.
+	 * @param context the back off context for this execution.
 	 */
 	public void maybeBackoff(Context context) {
-		long backoffTime = ChronoUnit.MILLIS.between(Instant.now(this.clock),
-				Instant.ofEpochMilli(context.dueTimestamp));
+		long backoffTime = context.dueTimestamp - getCurrentMillisFromClock();
 		if (backoffTime > 0) {
 			pauseConsumptionAndThrow(context, backoffTime);
 		}
 	}
 
-	private void pauseConsumptionAndThrow(Context context, Long timeToSleep) throws KafkaBackoffException {
+	private void pauseConsumptionAndThrow(Context context, Long backOffTime) throws KafkaBackoffException {
 		TopicPartition topicPartition = context.topicPartition;
 		getListenerContainerFromContext(context).pausePartition(topicPartition);
 		addBackoff(context, topicPartition);
 		throw new KafkaBackoffException(String.format("Partition %s from topic %s is not ready for consumption, " +
 				"backing off for approx. %s millis.", context.topicPartition.partition(),
-				context.topicPartition.topic(), timeToSleep),
+				context.topicPartition.topic(), backOffTime),
 				topicPartition, context.listenerId, context.dueTimestamp);
 	}
 
 	@Override
 	public void onApplicationEvent(ListenerContainerPartitionIdleEvent partitionIdleEvent) {
-		Context context = getBackoff(partitionIdleEvent.getTopicPartition());
-		if (context == null || isNotDue(context.dueTimestamp)) {
+		logger.debug(() -> String.format("partitionIdleEvent received in %s at %s. Partition: %s",
+				this.getClass().getSimpleName(), getCurrentMillisFromClock(), partitionIdleEvent.getTopicPartition()));
```

Comment:
Why do we need a class name in the logs? 
It could be derivatived from the log category, plus thread and sometime instance.
More over you use simple class name - no package. It is going confuse end-users.
I would just go without class at all...